File: Hill.py
import sys
import numpy as np


def cipher_encryption():
    msg = input("Enter message: ").upper()
    msg = msg.replace(" ", "")

    # if message length is odd number, append 0 at the end
    len_chk = 0
    if len(msg) % 2 != 0:
        msg += "0"
        len_chk = 1

    # msg to matrices
    row = 2
    col = int(len(msg)/2)
    msg2d = np.zeros((row, col), dtype=int)

    itr1 = 0
    itr2 = 0
    for i in range(len(msg)):
        if i % 2 == 0:
            msg2d[0][itr1] = int(ord(msg[i])-65)
            itr1 += 1
        else:
            msg2d[1][itr2] = int(ord(msg[i])-65)
            itr2 += 1
    # for

    key = input("Enter 4 letter Key String: ").upper()
    key = key.replace(" ", "")

    # key to 2x2
    key2d = np.zeros((2, 2), dtype=int)
    itr3 = 0
    for i in range(2):
        for j in range(2):
            key2d[i][j] = ord(key[itr3])-65
            itr3 += 1

    # checking validity of the key
    # finding determinant
    deter = key2d[0][0] * key2d[1][1] - key2d[0][1] * key2d[1][0]
    deter = deter % 26

    # finding multiplicative inverse
    mul_inv = -1
    for i in range(26):
        temp_inv = deter * i
        if temp_inv % 26 == 1:
            mul_inv = i
            break
        else:
            continue
    # for

    if mul_inv == -1:
        print("Invalid key")
        sys.exit()
    # if

    encryp_text = ""
    itr_count = int(len(msg)/2)
    if len_chk == 0:
        for i in range(itr_count):
            temp1 = msg2d[0][i] * key2d[0][0] + msg2d[1][i] * key2d[0][1]
            encryp_text += chr((temp1 % 26) + 65)
            temp2 = msg2d[0][i] * key2d[1][0] + msg2d[1][i] * key2d[1][1]
            encryp_text += chr((temp2 % 26) + 65)
        # for
    else:
        for i in range(itr_count-1):
            temp1 = msg2d[0][i] * key2d[0][0] + msg2d[1][i] * key2d[0][1]
            encryp_text += chr((temp1 % 26) + 65)
            temp2 = msg2d[0][i] * key2d[1][0] + msg2d[1][i] * key2d[1][1]
            encryp_text += chr((temp2 % 26) + 65)
        # for
    # if else

    print("Encrypted Text: {}".format(encryp_text))


def cipher_decryption():
    msg = input("Enter message: ").upper()
    msg = msg.replace(" ", "")

    # if message length is odd number, append 0 at the end
    len_chk = 0
    if len(msg) % 2 != 0:
        msg += "0"
        len_chk = 1

    # msg to matrices
    row = 2
    col = int(len(msg) / 2)
    msg2d = np.zeros((row, col), dtype=int)

    itr1 = 0
    itr2 = 0
    for i in range(len(msg)):
        if i % 2 == 0:
            msg2d[0][itr1] = int(ord(msg[i]) - 65)
            itr1 += 1
        else:
            msg2d[1][itr2] = int(ord(msg[i]) - 65)
            itr2 += 1
    # for

    key = input("Enter 4 letter Key String: ").upper()
    key = key.replace(" ", "")

    # key to 2x2
    key2d = np.zeros((2, 2), dtype=int)
    itr3 = 0
    for i in range(2):
        for j in range(2):
            key2d[i][j] = ord(key[itr3]) - 65
            itr3 += 1
    # for

    # finding determinant
    deter = key2d[0][0] * key2d[1][1] - key2d[0][1] * key2d[1][0]
    deter = deter % 26

    # finding multiplicative inverse
    mul_inv = -1
    for i in range(26):
        temp_inv = deter * i
        if temp_inv % 26 == 1:
            mul_inv = i
            break
        else:
            continue
    # for

    # adjugate matrix
    # swapping
    key2d[0][0], key2d[1][1] = key2d[1][1], key2d[0][0]

    # changing signs
    key2d[0][1] *= -1
    key2d[1][0] *= -1

    key2d[0][1] = key2d[0][1] % 26
    key2d[1][0] = key2d[1][0] % 26

    # multiplying multiplicative inverse with adjugate matrix
    for i in range(2):
        for j in range(2):
            key2d[i][j] *= mul_inv

    # modulo
    for i in range(2):
        for j in range(2):
            key2d[i][j] = key2d[i][j] % 26

    # cipher to plain
    decryp_text = ""
    itr_count = int(len(msg) / 2)
    if len_chk == 0:
        for i in range(itr_count):
            temp1 = msg2d[0][i] * key2d[0][0] + msg2d[1][i] * key2d[0][1]
            decryp_text += chr((temp1 % 26) + 65)
            temp2 = msg2d[0][i] * key2d[1][0] + msg2d[1][i] * key2d[1][1]
            decryp_text += chr((temp2 % 26) + 65)
            # for
    else:
        for i in range(itr_count - 1):
            temp1 = msg2d[0][i] * key2d[0][0] + msg2d[1][i] * key2d[0][1]
            decryp_text += chr((temp1 % 26) + 65)
            temp2 = msg2d[0][i] * key2d[1][0] + msg2d[1][i] * key2d[1][1]
            decryp_text += chr((temp2 % 26) + 65)
            # for
    # if else

    print("Decrypted Text: {}".format(decryp_text))


def main():
    choice = int(input("1. Encryption\n2. Decryption\nChoose(1,2): "))
    if choice == 1:
        print("---Encryption---")
        cipher_encryption()
    elif choice == 2:
        print("---Decryption---")
        cipher_decryption()
    else:
        print("Invalid Choice")

if __name__ == "__main__":
    main()

File: Hill.cpp
#include <iostream>
#include <regex>

using namespace std;

int moduloFunc(int a, int b){
    int result = a % b;
    if (result < 0){
        result += b;
    }
    return result;
}

void cipherEncryption(){
    string msg;
    cout << "Enter message: ";
    getline(cin, msg);

    // message to uppercase
    for (int i = 0; i < msg.length(); i++){
        msg[i] = toupper(msg[i]);
    }

    // removing white space from msg
    msg = regex_replace(msg, regex("\\s+"), "");

    // if msg.length %2 != 0 perform padding
    int lenChk = 0;
    if(msg.length()%2 != 0){
        msg += "0";
        lenChk = 1;
    }

    // message to 2x msg.length/2 matrix
    int msg2D[2][msg.length()/2];
    int itr1 = 0;
    int itr2 = 0;
    for (int i = 0; i < msg.length(); i++){
        if(i%2 == 0){
            msg2D[0][itr1] = msg[i] - 65;
            itr1++;
        } else {
            msg2D[1][itr2] = msg[i] - 65;
            itr2++;
        } //if-else
    } // for

//    // testing matrix
//    for(int i = 0; i < 2; i++){
//        for (int j = 0; j < msg.length()/2; j++){
//            cout << (char)(msg2D[i][j]+65) << " ";
//        }
//        cout << endl;
//    }

    cout << "Enter 4 letter key string: ";
    string key;
    getline(cin, key);

    // key to uppercase
    for (int i = 0; i < key.length(); i++){
        key[i] = toupper(key[i]);
    }

    // removing white space from key
    key = regex_replace(key, regex("\\s+"), "");

    // key to 2x2 matrix
    int key2D[2][2];
    int itr3 = 0;
    for (int i = 0; i < 2; i++){
        for (int j = 0; j<2; j++){
            key2D[i][j] = key[itr3]-65;
            itr3++;
        }
    }

    // checking validity of the key
    // finding determinant
    int deter = key2D[0][0] * key2D[1][1] - key2D[0][1] * key2D[1][0];
    deter = moduloFunc(deter, 26);

    // finding multiplicative inverse
    int mulInv = -1;
    for (int i=0; i<26; i++){
        int tempInv = deter * i;
        if (moduloFunc(tempInv, 26) == 1){
            mulInv = i;
            break;
        } else {
            continue;
        }
    } // for

    if (mulInv == -1){
        cout << "invalid key" << endl;
        exit(EXIT_FAILURE);
    }

    string encrypText = "";
    int itrCount = msg.length()/2;
    if (lenChk == 0){
        // if msg.length % 2 == 0
        for (int i = 0; i < itrCount; i++){
            int temp1 = msg2D[0][i] * key2D[0][0] + msg2D[1][i] * key2D[0][1];
            encrypText += (char)((temp1 % 26) + 65);
            int temp2 = msg2D[0][i] * key2D[1][0] + msg2D[1][i] * key2D[1][1];
            encrypText += (char)((temp2 % 26) + 65);
        }
    } else {
        // if msg.length % 2 == 0
        for (int i = 0; i < itrCount-1; i++){
            int temp1 = msg2D[0][i] * key2D[0][0] + msg2D[1][i] * key2D[0][1];
            encrypText += (char)((temp1 % 26) + 65);
            int temp2 = msg2D[0][i] * key2D[1][0] + msg2D[1][i] * key2D[1][1];
            encrypText += (char)((temp2 % 26) + 65);
        }
    }

    cout << endl << "Encrypted text: " << encrypText << endl;
}

void cipherDecryption(){
    string msg;
    cout << "Enter message: ";
    getline(cin, msg);

    // message to uppercase
    for (int i = 0; i < msg.length(); i++){
        msg[i] = toupper(msg[i]);
    }

    // removing white space from msg
    msg = regex_replace(msg, regex("\\s+"), "");

    // if msg.length %2 != 0 perform padding
    int lenChk = 0;
    if(msg.length()%2 != 0){
        msg += "0";
        lenChk = 1;
    }

    // message to 2x msg.length/2 matrix
    int msg2D[2][msg.length()/2];
    int itr1 = 0;
    int itr2 = 0;
    for (int i = 0; i < msg.length(); i++){
        if(i%2 == 0){
            msg2D[0][itr1] = msg[i] - 65;
            itr1++;
        } else {
            msg2D[1][itr2] = msg[i] - 65;
            itr2++;
        } //if-else
    } // for

//    // testing matrix
//    for(int i = 0; i < 2; i++){
//        for (int j = 0; j < msg.length()/2; j++){
//            cout << (char)(msg2D[i][j]+65) << " ";
//        }
//        cout << endl;
//    }

    cout << "Enter 4 letter key string: ";
    string key;
    getline(cin, key);

    // key to uppercase
    for (int i = 0; i < key.length(); i++){
        key[i] = toupper(key[i]);
    }

    // removing white space from key
    key = regex_replace(key, regex("\\s+"), "");

    // key to 2x2 matrix
    int key2D[2][2];
    int itr3 = 0;
    for (int i = 0; i < 2; i++){
        for (int j = 0; j<2; j++){
            key2D[i][j] = key[itr3]-65;
            itr3++;
        }
    }

    // finding determinant
    int deter = key2D[0][0] * key2D[1][1] - key2D[0][1] * key2D[1][0];
    deter = moduloFunc(deter, 26);

    // finding multiplicative inverse
    int mulInv = -1;
    for (int i=0; i<26; i++){
        int tempInv = deter * i;
        if (moduloFunc(tempInv, 26) == 1){
            mulInv = i;
            break;
        } else {
            continue;
        }
    } // for


    // adjugate matrix
    //swapping
    swap(key2D[0][0],key2D[1][1]);

    // changing signs
    key2D[0][1] *= -1;
    key2D[1][0] *= -1;

    key2D[0][1] = moduloFunc(key2D[0][1], 26);
    key2D[1][0] = moduloFunc(key2D[1][0], 26);

    // multiplying multiplicative inverse with adjugate matrix
    for (int i = 0; i < 2; i++){
        for (int j = 0; j < 2; j++){
            key2D[i][j] *= mulInv;
        }
    }
    for (int i=0; i<2;i++){
        for(int j =0; j <2; j++){
            key2D[i][j] = moduloFunc(key2D[i][j], 26);
        }
    }

    // ciphertext to plain text
    string decrypText = "";
    int itrCount = msg.length()/2;
    if (lenChk == 0){
        // if msg.length % 2 == 0
        for (int i = 0; i < itrCount; i++){
            int temp1 = msg2D[0][i] * key2D[0][0] + msg2D[1][i] * key2D[0][1];
            decrypText += (char)((temp1 % 26) + 65);
            int temp2 = msg2D[0][i] * key2D[1][0] + msg2D[1][i] * key2D[1][1];
            decrypText += (char)((temp2 % 26) + 65);
        }
    } else {
        // if msg.length % 2 == 0
        for (int i = 0; i < itrCount-1; i++){
            int temp1 = msg2D[0][i] * key2D[0][0] + msg2D[1][i] * key2D[0][1];
            decrypText += (char)((temp1 % 26) + 65);
            int temp2 = msg2D[0][i] * key2D[1][0] + msg2D[1][i] * key2D[1][1];
            decrypText += (char)((temp2 % 26) + 65);
        }
    }

    cout << endl << "Decrypted text: " << decrypText << endl;

}

int main()
{
    cout << "1. Encryption\n2. Decryption\nChoose(1,2): ";
    int choice;
    cin >> choice;
    cin.ignore();

    if (choice == 1){
        cout << endl << "---Encryption---" << endl;
        cipherEncryption();
    } else if (choice == 2){
        cout << endl << "---Decryption---" << endl;
        cipherDecryption();
    } else {
        cout << endl << "Wrong choice" << endl;
    }

    return 0;
}
